File: snippets/cpp/tree/lca.cpp
struct LCA {
    int n, lg; vector<vi> up, g; vi dep;
    LCA(vector<vi> &g, int root = 0) : n(sza(g)), lg(__lg(n) + 1), up(lg, vi(n)), g(g), dep(n) {
        dfs(root, root);
        for (int i = 1; i < lg; i++)
            for (int u = 0; u < n; u++)
                up[i][u] = up[i-1][up[i-1][u]];
    }
    void dfs(int u, int p) {
        up[0][u] = p;
        for (int v : g[u])
            if (v != p) dep[v] = dep[u] + 1, dfs(v, u);
    }
    int lca(int u, int v) {
        if (dep[u] < dep[v]) swap(u, v);
        int diff = dep[u] - dep[v];
        for (int i = 0; i < lg; i++)
            if (diff >> i & 1) u = up[i][u];
        if (u == v) return u;
        for (int i = lg - 1; i >= 0; i--)
            if (up[i][u] != up[i][v])
                u = up[i][u], v = up[i][v];
        return up[0][u];
    }
    int dist(int u, int v) { return dep[u] + dep[v] - 2 * dep[lca(u, v)]; }
};

File: snippets/rust/tree/lca.rs
// Binary-lifting LCA. Uses iterative BFS to avoid stack overflow on deep trees.
// `g` is an undirected adjacency list. Pass the root index.
struct LCA {
    lg: usize,
    up: Vec<Vec<usize>>,
    dep: Vec<usize>,
}
impl LCA {
    fn new(g: &Vec<Vec<usize>>, root: usize) -> Self {
        let n = g.len();
        let lg = (usize::BITS - (n.max(1) - 1).leading_zeros()).max(1) as usize;
        let mut up = vec![vec![root; n]; lg];
        let mut dep = vec![0usize; n];
        let mut parent = vec![root; n];
        let mut visited = vec![false; n];
        let mut q = std::collections::VecDeque::new();
        q.push_back(root);
        visited[root] = true;
        while let Some(u) = q.pop_front() {
            for &v in &g[u] {
                if !visited[v] {
                    visited[v] = true;
                    parent[v] = u;
                    dep[v] = dep[u] + 1;
                    q.push_back(v);
                }
            }
        }
        for u in 0..n { up[0][u] = parent[u]; }
        for i in 1..lg {
            for u in 0..n { up[i][u] = up[i - 1][up[i - 1][u]]; }
        }
        Self { lg, up, dep }
    }
    fn lca(&self, mut u: usize, mut v: usize) -> usize {
        if self.dep[u] < self.dep[v] { std::mem::swap(&mut u, &mut v); }
        let diff = self.dep[u] - self.dep[v];
        for i in 0..self.lg {
            if (diff >> i) & 1 == 1 { u = self.up[i][u]; }
        }
        if u == v { return u; }
        for i in (0..self.lg).rev() {
            if self.up[i][u] != self.up[i][v] {
                u = self.up[i][u];
                v = self.up[i][v];
            }
        }
        self.up[0][u]
    }
    fn dist(&self, u: usize, v: usize) -> usize {
        let w = self.lca(u, v);
        self.dep[u] + self.dep[v] - 2 * self.dep[w]
    }
}
